File: mobile/apps/photos/plugins/ente_photos_platform/ios/Flutter/ProcessLockChannelAdapter.swift
@preconcurrency import Flutter
import Foundation

@MainActor
final class ProcessLockChannelAdapter: NSObject {
    private let methodChannel: FlutterMethodChannel
    private let instanceId = UUID().uuidString
    private var isAttached = true

    init(registrar: FlutterPluginRegistrar) {
        methodChannel = FlutterMethodChannel(
            name: Self.methodChannelName,
            binaryMessenger: registrar.messenger()
        )
        super.init()
        methodChannel.setMethodCallHandler { [weak self] call, result in
            self?.handle(call, result: result)
        }
    }

    func detach() {
        guard isAttached else { return }
        isAttached = false
        ProcessLockRegistry.shared.releaseAll(instanceId: instanceId)
        methodChannel.setMethodCallHandler(nil)
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard isAttached else {
            result(
                FlutterError(code: "detached", message: "adapter detached", details: nil)
            )
            return
        }
        switch call.method {
        case "processLock.tryAcquire":
            handleTryAcquire(call, result: result)
        case "processLock.release":
            handleRelease(call, result: result)
        case "processLock.state":
            handleState(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handleTryAcquire(_ call: FlutterMethodCall, result: FlutterResult) {
        guard let args = call.arguments as? [String: Any],
            let name = args["name"] as? String, !name.isEmpty,
            let origin = args["origin"] as? String, !origin.isEmpty,
            let operation = args["operation"] as? String, !operation.isEmpty
        else {
            result(
                FlutterError(
                    code: "invalidArguments",
                    message: "name, origin and operation are required",
                    details: nil
                )
            )
            return
        }
        result(
            ProcessLockRegistry.shared.tryAcquire(
                name: name,
                instanceId: instanceId,
                origin: origin,
                operation: operation
            )
        )
    }

    private func handleRelease(_ call: FlutterMethodCall, result: FlutterResult) {
        guard let args = call.arguments as? [String: Any],
            let name = args["name"] as? String, !name.isEmpty
        else {
            result(
                FlutterError(
                    code: "invalidArguments",
                    message: "name is required",
                    details: nil
                )
            )
            return
        }
        result(ProcessLockRegistry.shared.release(name: name, instanceId: instanceId))
    }

    private func handleState(_ call: FlutterMethodCall, result: FlutterResult) {
        guard let args = call.arguments as? [String: Any],
            let name = args["name"] as? String, !name.isEmpty
        else {
            result(
                FlutterError(
                    code: "invalidArguments",
                    message: "name is required",
                    details: nil
                )
            )
            return
        }
        result(ProcessLockRegistry.shared.state(name: name))
    }

    private static let methodChannelName = "io.ente.photos.platform/process_lock"
}

File: mobile/apps/photos/plugins/ente_photos_platform/android/src/main/kotlin/io/ente/photos/platform/flutter/ProcessLockChannelAdapter.kt
package io.ente.photos.platform.flutter

import io.ente.photos.platform.processlock.ProcessLockRegistry
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.UUID

internal class ProcessLockChannelAdapter : MethodChannel.MethodCallHandler {
    private lateinit var methodChannel: MethodChannel
    private val instanceId = UUID.randomUUID().toString()

    fun attach(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        methodChannel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "processLock.tryAcquire" -> {
                val name = call.argument<String>("name")
                val origin = call.argument<String>("origin")
                val operation = call.argument<String>("operation")
                if (name.isNullOrEmpty() || origin.isNullOrEmpty() || operation.isNullOrEmpty()) {
                    result.error(
                        "invalidArguments",
                        "name, origin and operation are required",
                        null,
                    )
                } else {
                    result.success(
                        ProcessLockRegistry.tryAcquire(name, instanceId, origin, operation),
                    )
                }
            }
            "processLock.release" -> {
                val name = call.argument<String>("name")
                if (name.isNullOrEmpty()) {
                    result.error("invalidArguments", "name is required", null)
                } else {
                    result.success(ProcessLockRegistry.release(name, instanceId))
                }
            }
            "processLock.state" -> {
                val name = call.argument<String>("name")
                if (name.isNullOrEmpty()) {
                    result.error("invalidArguments", "name is required", null)
                } else {
                    result.success(ProcessLockRegistry.state(name))
                }
            }
            else -> result.notImplemented()
        }
    }

    fun detach() {
        ProcessLockRegistry.releaseAllForInstance(instanceId)
        methodChannel.setMethodCallHandler(null)
    }

    private companion object {
        const val METHOD_CHANNEL = "io.ente.photos.platform/process_lock"
    }
}
